checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,363 @@
+        AnnotationMirror writtenMCAnno = type.getAnnotationInHierarchy(this.TOP);
+        if (writtenMCAnno != null
+            && !this.getQualifierHierarchy().isSubtype(inheritedMCAnno, writtenMCAnno)) {
+          if (!elementsIssuedInconsistentMustCallSubtypeErrors.contains(elt)
+              && !this.checker.shouldSkipUses(elt)) {
+            checker.reportError(
+                elt,
+                "inconsistent.mustcall.subtype",
+                elt.getSimpleName(),
+                writtenMCAnno,
+                inheritableMustCall);
+            elementsIssuedInconsistentMustCallSubtypeErrors.add(elt);
+          }
+        } else {
+          type.replaceAnnotation(inheritedMCAnno);
+        }
+      }
+    }
+    return type;
+  }
+
+  /**
+   * Creates a {@link MustCall} annotation whose values are the given strings.
+   *
+   * @param val the methods that should be called
+   * @return an annotation indicating that the given methods should be called
+   */
+  public AnnotationMirror createMustCall(final String... val) {
+    AnnotationBuilder builder = new AnnotationBuilder(processingEnv, MustCall.class);
+    Arrays.sort(val);
+    builder.setValue("value", val);
+    return builder.build();
+  }
+
+  @Override
+  public QualifierHierarchy createQualifierHierarchy() {
+    return new SubtypeIsSubsetQualifierHierarchy(
+        this.getSupportedTypeQualifiers(), this.getProcessingEnv());
+  }
+
+  /**
+   * Fetches the store from the results of dataflow, for either block (if noSuccInfo is true) or
+   * succ (if noSuccInfo is false).
+   *
+   * @param noSuccInfo whether to use the store for the block itself or its successor, succ
+   * @param block a block
+   * @param succ block's successor
+   * @return the appropriate CFStore, populated with MustCall annotations, from the results of
+   *     running dataflow
+   */
+  public CFStore getStoreForBlock(boolean noSuccInfo, Block block, Block succ) {
+    return noSuccInfo ? flowResult.getStoreAfter(block) : flowResult.getStoreBefore(succ);
+  }
+
+  /**
+   * The TreeAnnotator for the MustCall type system. This tree annotator treats non-owning method
+   * parameters as bottom, regardless of their declared type, when they appear in the body of the
+   * method. Doing so is safe because being non-owning means, by definition, that their must-call
+   * obligations are only relevant at the call site.
+   */
+  private class MustCallTreeAnnotator extends TreeAnnotator {
+    /**
+     * Create a MustCallTreeAnnotator
+     *
+     * @param mustCallAnnotatedTypeFactory the type factory
+     */
+    public MustCallTreeAnnotator(MustCallAnnotatedTypeFactory mustCallAnnotatedTypeFactory) {
+      super(mustCallAnnotatedTypeFactory);
+    }
+
+    // When they appear in the body of a method or constructor, treat non-owning parameters
+    // as bottom regardless of their declared type.
+    @Override
+    public Void visitIdentifier(IdentifierTree node, AnnotatedTypeMirror type) {
```

Comment:
Because doing so would require directly modifying one of the caches, which is error-prone and relies on an implementation detail of the type factory. I always try to avoid doing so. The only other way that I know to accomplish this task easily would be to add logic to `getComputedTypeAnnotations`, which is worse (it runs when computing every type, rather than just the type of an identifier).

In particular, I tried to override `visitVariable` to directly change the type of the parameter, but the following unexpected diagnostics are issued:
```
    NonOwningPolyInteraction.java:9: error: (assignment.type.incompatible)
    NonOwningPolyInteraction.java:21: error: (assignment.type.incompatible)
    NonOwningPolyInteraction.java:26: error: (assignment.type.incompatible)
    NonOwningPolyInteraction.java:40: error: (assignment.type.incompatible)
```